crutest/src/main.rs

```diff
@@ -2185,11 +2351,149 @@ async fn perf_workload(
     }
     Ok(())
 }
+
+/// Prints a pleasant summary of the given region
+fn print_region_description(ri: &RegionInfo, encrypted: bool) {
+    println!("region info:");
+    println!("  block size:      {} bytes", ri.block_size);
+    println!("  blocks / extent: {}", ri.extent_size.value);
+    println!(
+        "  extent size:     {}",
+        human_bytes((ri.block_size * ri.extent_size.value) as f64)
+    );
+    println!(
+        "  extent count:    {}",
+        ri.total_blocks as u64 / ri.extent_size.value
+    );
+    println!("  total blocks:    {}", ri.total_blocks);
+    println!("  total size:      {}", human_bytes(ri.total_size as f64));
+    println!(
+        "  encryption:      {}",
+        if encrypted { "yes" } else { "no" }
+    );
+}
+
+async fn rand_read_write_workload(
+    guest: &Arc<Guest>,
+    ri: &mut RegionInfo,
+    cfg: RandReadWriteConfig,
+) -> Result<()> {
+    // Before we start, make sure the work queues are empty.
+    loop {
+        let wc = guest.query_work_queue().await?;
+        if wc.up_count + wc.ds_count == 0 {
+            break;
+        }
+        tokio::time::sleep(Duration::from_secs(2)).await;
+    }
+
+    let desc = match cfg.mode {
+        RandReadWriteMode::Read => "read",
+        RandReadWriteMode::Write => "write",
+    };
+
+    if cfg.fill {
+        println!("filling region before {desc}");
+        fill_workload(guest, ri, true).await?;
+    }
+
+    if cfg.blocks_per_io > ri.total_blocks {
+        bail!("too many blocks per IO; can't exceed {}", ri.total_blocks);
+    }
+
+    println!(
+        "\n-----------------------------------\
+        \nrandom {desc} with {} chunks ({} block{})",
+        human_bytes((cfg.blocks_per_io as u64 * ri.block_size) as f64),
+        cfg.blocks_per_io,
+        if cfg.blocks_per_io > 1 { "s" } else { "" },
+    );
+    print_region_description(ri, cfg.encrypted);
+
+    let stop = Arc::new(AtomicBool::new(false));
+    let byte_count = Arc::new(AtomicUsize::new(0));
+
+    let block_size = ri.block_size as usize;
+    let total_blocks = ri.total_blocks;
+    for _ in 0..cfg.io_depth {
+        let stop = stop.clone();
+        let guest = guest.clone();
+        let byte_count = byte_count.clone();
+        tokio::spawn(async move {
+            match cfg.mode {
+                RandReadWriteMode::Write => {
+                    let mut buf = BytesMut::new();
+                    buf.resize(cfg.blocks_per_io * block_size, 0u8);
+                    let mut rng = rand_chacha::ChaCha8Rng::from_entropy();
+                    rng.fill_bytes(&mut buf);
```

Comment:
If you generate and fill the buffer inside the loop, won't that get counted in the
performance results?  I don't know if we care really that much, or how much time we
actually spend in the allocator.

Reply:
This is outside the IO loop, which begins at `while !stop.load(..)` 😉

Reply:
Ah yeah, got it.